--- ../src-base/minecraft/net/minecraft/entity/player/EntityPlayer.java
+++ ../src-work/minecraft/net/minecraft/entity/player/EntityPlayer.java
@@ -1,5 +1,7 @@
 package net.minecraft.entity.player;
 
+import cpw.mods.fml.common.FMLCommonHandler;
+import cpw.mods.fml.common.network.FMLNetworkHandler;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.util.Iterator;
@@ -52,9 +54,16 @@
 import net.minecraft.world.EnumGameType;
 import net.minecraft.world.World;
 import net.minecraft.world.chunk.IChunkProvider;
+import net.minecraftforge.common.ForgeHooks;
+import net.minecraftforge.common.ISpecialArmor;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.ForgeEventFactory;
+import net.minecraftforge.event.entity.player.*;
 
 public abstract class EntityPlayer extends EntityLiving implements ICommandSender
 {
+    public static final String PERSISTED_NBT_TAG = "PlayerPersisted";
+    
     public InventoryPlayer field_71071_by = new InventoryPlayer(this);
     private InventoryEnderChest field_71078_a = new InventoryEnderChest();
     public Container field_71069_bz;
@@ -173,12 +182,14 @@
 
     public void func_70071_h_()
     {
+        FMLCommonHandler.instance().onPlayerPreTick(this);
         if (this.field_71074_e != null)
         {
             ItemStack var1 = this.field_71071_by.func_70448_g();
 
             if (var1 == this.field_71074_e)
             {
+                field_71074_e.func_77973_b().onUsingItemTick(field_71074_e, this, field_71072_f);
                 if (this.field_71072_f <= 25 && this.field_71072_f % 4 == 0)
                 {
                     this.func_71010_c(var1, 5);
@@ -296,6 +307,7 @@
         {
             this.field_71100_bB.func_75118_a(this);
         }
+        FMLCommonHandler.instance().onPlayerPostTick(this);
     }
 
     public int func_82145_z()
@@ -395,11 +407,11 @@
         this.field_71109_bG = 0.0F;
         this.func_71015_k(this.field_70165_t - var1, this.field_70163_u - var3, this.field_70161_v - var5);
 
-        if (this.field_70154_o instanceof EntityPig)
+        if (this.field_70154_o instanceof EntityLiving && ((EntityLiving)field_70154_o).shouldRiderFaceForward(this))
         {
             this.field_70125_A = var8;
             this.field_70177_z = var7;
-            this.field_70761_aq = ((EntityPig)this.field_70154_o).field_70761_aq;
+            this.field_70761_aq = ((EntityLiving)this.field_70154_o).field_70761_aq;
         }
     }
 
@@ -509,6 +521,9 @@
         this.func_70105_a(0.2F, 0.2F);
         this.func_70107_b(this.field_70165_t, this.field_70163_u, this.field_70161_v);
         this.field_70181_x = 0.10000000149011612D;
+        
+        captureDrops = true;
+        capturedDrops.clear();
 
         if (this.field_71092_bJ.equals("Notch"))
         {
@@ -519,6 +534,20 @@
         {
             this.field_71071_by.func_70436_m();
         }
+        
+        captureDrops = false;
+        
+        if (!field_70170_p.field_72995_K)
+        {
+            PlayerDropsEvent event = new PlayerDropsEvent(this, p_70645_1_, capturedDrops, field_70718_bc > 0);
+            if (!MinecraftForge.EVENT_BUS.post(event))
+            {
+                for (EntityItem item : capturedDrops)
+                {
+                    func_71012_a(item);
+                }
+            }
+        }
 
         if (p_70645_1_ != null)
         {
@@ -550,12 +579,25 @@
 
     public EntityItem func_71040_bB(boolean p_71040_1_)
     {
-        return this.func_71019_a(this.field_71071_by.func_70298_a(this.field_71071_by.field_70461_c, p_71040_1_ && this.field_71071_by.func_70448_g() != null ? this.field_71071_by.func_70448_g().field_77994_a : 1), false);
+        ItemStack stack = field_71071_by.func_70448_g();
+        
+        if (stack == null)
+        {
+            return null;
+        }
+        
+        if (stack.func_77973_b().onDroppedByPlayer(stack, this))
+        {
+            int count = p_71040_1_ && this.field_71071_by.func_70448_g() != null ? this.field_71071_by.func_70448_g().field_77994_a : 1;
+            return ForgeHooks.onPlayerTossEvent(this, field_71071_by.func_70298_a(field_71071_by.field_70461_c, count));
+        }
+        
+        return null;
     }
 
     public EntityItem func_71021_b(ItemStack p_71021_1_)
     {
-        return this.func_71019_a(p_71021_1_, false);
+        return ForgeHooks.onPlayerTossEvent(this, p_71021_1_);
     }
 
     public EntityItem func_71019_a(ItemStack p_71019_1_, boolean p_71019_2_)
@@ -601,12 +643,29 @@
 
     public void func_71012_a(EntityItem p_71012_1_)
     {
-        this.field_70170_p.func_72838_d(p_71012_1_);
+        if (captureDrops)
+        {
+            capturedDrops.add(p_71012_1_);
+        }
+        else
+        {
+            this.field_70170_p.func_72838_d(p_71012_1_);
+        }
     }
 
+    /**
+     * Deprecated in favor of the more sensitive version
+     */
+    @Deprecated
     public float func_71055_a(Block p_71055_1_)
     {
-        float var2 = this.field_71071_by.func_70438_a(p_71055_1_);
+        return getCurrentPlayerStrVsBlock(p_71055_1_, 0);
+    }
+    
+    public float getCurrentPlayerStrVsBlock(Block par1Block, int meta)
+    {
+        ItemStack stack = field_71071_by.func_70448_g();
+        float var2 = (stack == null ? 1.0F : stack.func_77973_b().getStrVsBlock(stack, par1Block, meta));
         int var3 = EnchantmentHelper.func_77509_b(this);
         ItemStack var4 = this.field_71071_by.func_70448_g();
 
@@ -614,7 +673,9 @@
         {
             float var5 = (float)(var3 * var3 + 1);
 
-            if (!var4.func_77987_b(p_71055_1_) && var2 <= 1.0F)
+            boolean canHarvest = ForgeHooks.canToolHarvestBlock(par1Block, meta, var4);
+            
+            if (!canHarvest && var2 <= 1.0F)
             {
                 var2 += var5 * 0.08F;
             }
@@ -644,12 +705,13 @@
             var2 /= 5.0F;
         }
 
-        return var2;
+        var2 = ForgeEventFactory.getBreakSpeed(this, par1Block, meta, var2);
+        return (var2 < 0 ? 0 : var2);
     }
 
     public boolean func_71062_b(Block p_71062_1_)
     {
-        return this.field_71071_by.func_70454_b(p_71062_1_);
+        return ForgeEventFactory.doPlayerHarvestCheck(this, p_71062_1_, field_71071_by.func_70454_b(p_71062_1_));
     }
 
     public void func_70037_a(NBTTagCompound p_70037_1_)
@@ -685,6 +747,10 @@
             NBTTagList var3 = p_70037_1_.func_74761_m("EnderItems");
             this.field_71078_a.func_70486_a(var3);
         }
+        
+        //Disable vanilla mob armor/item pickup, players can already pickup items 
+        //and it causes issues with overwriting items, dropping items, etc..
+        this.field_82172_bs = false;
     }
 
     public void func_70014_b(NBTTagCompound p_70014_1_)
@@ -906,12 +972,22 @@
     {
         if (!this.func_85032_ar())
         {
+            p_70665_2_ = ForgeHooks.onLivingHurt(this, p_70665_1_, p_70665_2_);
+            if (p_70665_2_ <= 0)
+            {
+                return;
+            }
+            
             if (!p_70665_1_.func_76363_c() && this.func_70632_aY())
             {
                 p_70665_2_ = 1 + p_70665_2_ >> 1;
             }
 
-            p_70665_2_ = this.func_70655_b(p_70665_1_, p_70665_2_);
+            p_70665_2_ = ISpecialArmor.ArmorProperties.ApplyArmor(this, field_71071_by.field_70460_b, p_70665_1_, p_70665_2_);
+            if (p_70665_2_ <= 0)
+            {
+                return;
+            }
             p_70665_2_ = this.func_70672_c(p_70665_1_, p_70665_2_);
             this.func_71020_j(p_70665_1_.func_76345_d());
             this.field_70734_aK -= p_70665_2_;
@@ -948,6 +1024,10 @@
 
     public boolean func_70998_m(Entity p_70998_1_)
     {
+        if (MinecraftForge.EVENT_BUS.post(new EntityInteractEvent(this, p_70998_1_)))
+        {
+            return false;
+        }
         if (p_70998_1_.func_70085_c(this))
         {
             return true;
@@ -985,7 +1065,9 @@
 
     public void func_71028_bD()
     {
+        ItemStack orig = func_71045_bC();
         this.field_71071_by.func_70299_a(this.field_71071_by.field_70461_c, (ItemStack)null);
+        MinecraftForge.EVENT_BUS.post(new PlayerDestroyItemEvent(this, orig));
     }
 
     public double func_70033_W()
@@ -995,6 +1077,15 @@
 
     public void func_71059_n(Entity p_71059_1_)
     {
+        if (MinecraftForge.EVENT_BUS.post(new AttackEntityEvent(this, p_71059_1_)))
+        {
+            return;
+        }
+        ItemStack stack = func_71045_bC();
+        if (stack != null && stack.func_77973_b().onLeftClickEntity(stack, this, p_71059_1_))
+        {
+            return;
+        }
         if (p_71059_1_.func_70075_an())
         {
             if (!p_71059_1_.func_85031_j(this))
@@ -1152,6 +1243,12 @@
 
     public EnumStatus func_71018_a(int p_71018_1_, int p_71018_2_, int p_71018_3_)
     {
+        PlayerSleepInBedEvent event = new PlayerSleepInBedEvent(this, p_71018_1_, p_71018_2_, p_71018_3_);
+        MinecraftForge.EVENT_BUS.post(event);
+        if (event.result != null)
+        {
+            return event.result;
+        }
         if (!this.field_70170_p.field_72995_K)
         {
             if (this.func_70608_bn() || !this.func_70089_S())
@@ -1191,6 +1288,11 @@
         {
             int var9 = this.field_70170_p.func_72805_g(p_71018_1_, p_71018_2_, p_71018_3_);
             int var5 = BlockBed.func_72217_d(var9);
+            Block block = Block.field_71973_m[field_70170_p.func_72798_a(p_71018_1_, p_71018_2_, p_71018_3_)];
+            if (block != null)
+            {
+                var5 = block.getBedDirection(field_70170_p, p_71018_1_, p_71018_2_, p_71018_3_);
+            }
             float var10 = 0.5F;
             float var7 = 0.5F;
 
@@ -1258,10 +1360,12 @@
         ChunkCoordinates var4 = this.field_71081_bT;
         ChunkCoordinates var5 = this.field_71081_bT;
 
-        if (var4 != null && this.field_70170_p.func_72798_a(var4.field_71574_a, var4.field_71572_b, var4.field_71573_c) == Block.field_71959_S.field_71990_ca)
+        Block block = (var4 == null ? null : Block.field_71973_m[field_70170_p.func_72798_a(var4.field_71574_a, var4.field_71572_b, var4.field_71573_c)]);
+        
+        if (var4 != null && block != null && block.isBed(field_70170_p, var4.field_71574_a, var4.field_71572_b, var4.field_71573_c, this))
         {
-            BlockBed.func_72228_a(this.field_70170_p, var4.field_71574_a, var4.field_71572_b, var4.field_71573_c, false);
-            var5 = BlockBed.func_72226_b(this.field_70170_p, var4.field_71574_a, var4.field_71572_b, var4.field_71573_c, 0);
+            block.setBedOccupied(this.field_70170_p, var4.field_71574_a, var4.field_71572_b, var4.field_71573_c, this, false);
+            var5 = block.getBedSpawnPosition(field_70170_p, var4.field_71574_a, var4.field_71572_b, var4.field_71573_c, this);
 
             if (var5 == null)
             {
@@ -1295,7 +1399,9 @@
 
     private boolean func_71065_l()
     {
-        return this.field_70170_p.func_72798_a(this.field_71081_bT.field_71574_a, this.field_71081_bT.field_71572_b, this.field_71081_bT.field_71573_c) == Block.field_71959_S.field_71990_ca;
+        ChunkCoordinates c = field_71081_bT;
+        int blockID = field_70170_p.func_72798_a(c.field_71574_a, c.field_71572_b, c.field_71573_c);
+        return Block.field_71973_m[blockID] != null && Block.field_71973_m[blockID].isBed(field_70170_p, c.field_71574_a, c.field_71572_b, c.field_71573_c, this);
     }
 
     public static ChunkCoordinates func_71056_a(World p_71056_0_, ChunkCoordinates p_71056_1_, boolean p_71056_2_)
@@ -1306,9 +1412,12 @@
         var3.func_73158_c(p_71056_1_.field_71574_a - 3 >> 4, p_71056_1_.field_71573_c + 3 >> 4);
         var3.func_73158_c(p_71056_1_.field_71574_a + 3 >> 4, p_71056_1_.field_71573_c + 3 >> 4);
 
-        if (p_71056_0_.func_72798_a(p_71056_1_.field_71574_a, p_71056_1_.field_71572_b, p_71056_1_.field_71573_c) == Block.field_71959_S.field_71990_ca)
+        ChunkCoordinates c = p_71056_1_;
+        Block block = Block.field_71973_m[p_71056_0_.func_72798_a(c.field_71574_a, c.field_71572_b, c.field_71573_c)];
+        
+        if (block != null && block.isBed(p_71056_0_, c.field_71574_a, c.field_71572_b, c.field_71573_c, null))
         {
-            ChunkCoordinates var8 = BlockBed.func_72226_b(p_71056_0_, p_71056_1_.field_71574_a, p_71056_1_.field_71572_b, p_71056_1_.field_71573_c, 0);
+            ChunkCoordinates var8 = block.getBedSpawnPosition(p_71056_0_, c.field_71574_a, c.field_71572_b, c.field_71573_c, null);
             return var8;
         }
         else
@@ -1326,8 +1435,11 @@
     {
         if (this.field_71081_bT != null)
         {
-            int var1 = this.field_70170_p.func_72805_g(this.field_71081_bT.field_71574_a, this.field_71081_bT.field_71572_b, this.field_71081_bT.field_71573_c);
-            int var2 = BlockBed.func_72217_d(var1);
+            int x = field_71081_bT.field_71574_a;
+            int y = field_71081_bT.field_71572_b;
+            int z = field_71081_bT.field_71573_c;
+            Block block = Block.field_71973_m[field_70170_p.func_72798_a(x, y, z)];
+            int var2 = (block == null ? 0 : block.getBedDirection(field_70170_p, x, y, z));
 
             switch (var2)
             {
@@ -1355,7 +1467,6 @@
         return this.field_71083_bS && this.field_71076_b >= 100;
     }
 
-    @SideOnly(Side.CLIENT)
     public int func_71060_bI()
     {
         return this.field_71076_b;
@@ -1594,7 +1705,7 @@
         {
             if (p_70620_1_.func_77973_b().func_77623_v())
             {
-                return p_70620_1_.func_77973_b().func_77618_c(p_70620_1_.func_77960_j(), p_70620_2_);
+                return p_70620_1_.func_77973_b().getIconIndex(p_70620_1_, p_70620_2_);
             }
 
             if (this.field_71074_e != null && p_70620_1_.field_77993_c == Item.field_77707_k.field_77779_bT)
@@ -1616,6 +1727,7 @@
                     return 101;
                 }
             }
+            var3 = p_70620_1_.func_77973_b().getIconIndex(p_70620_1_, p_70620_2_, this, field_71074_e, field_71072_f);
         }
 
         return var3;
@@ -1802,6 +1914,14 @@
         }
 
         this.field_71078_a = p_71049_1_.field_71078_a;
+        
+        //Copy over a section of the Entity Data from the old player.
+        //Allows mods to specify data that persists after players respawn.
+        NBTTagCompound old = p_71049_1_.getEntityData();
+        if (old.func_74764_b(PERSISTED_NBT_TAG))
+        {
+            getEntityData().func_74766_a(PERSISTED_NBT_TAG, old.func_74775_l(PERSISTED_NBT_TAG));
+        }
     }
 
     protected boolean func_70041_e_()
@@ -1849,7 +1969,14 @@
 
     public void func_70062_b(int p_70062_1_, ItemStack p_70062_2_)
     {
-        this.field_71071_by.field_70460_b[p_70062_1_] = p_70062_2_;
+        if (p_70062_1_ == 0)
+        {
+            this.field_71071_by.field_70462_a[this.field_71071_by.field_70461_c] = p_70062_2_;
+        }
+        else
+        {
+            this.field_71071_by.field_70460_b[p_70062_1_ - 1] = p_70062_2_;
+        }
     }
 
     public ItemStack[] func_70035_c()
@@ -1861,5 +1988,10 @@
     public boolean func_82238_cc()
     {
         return this.func_82241_s(1);
+    }
+    
+    public void openGui(Object mod, int modGuiId, World world, int x, int y, int z)
+    {
+        FMLNetworkHandler.openGui(this, mod, modGuiId, world, x, y, z);
     }
 }
