--- a/net/minecraft/entity/player/EntityPlayerMP.java
+++ b/net/minecraft/entity/player/EntityPlayerMP.java
@@ -9,6 +9,7 @@
 import java.util.List;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.IMerchant;
+import net.minecraft.entity.item.EntityItem;
 import net.minecraft.entity.projectile.EntityArrow;
 import net.minecraft.inventory.Container;
 import net.minecraft.inventory.ContainerBeacon;
@@ -73,6 +74,10 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
 import net.minecraft.world.chunk.Chunk;
+import net.minecraftforge.common.ForgeHooks;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.entity.player.PlayerDropsEvent;
+import net.minecraftforge.event.world.ChunkWatchEvent;
 
 public class EntityPlayerMP extends EntityPlayer implements ICrafting
 {
@@ -103,19 +108,11 @@
         p_i3396_4_.field_73090_b = this;
         this.field_71134_c = p_i3396_4_;
         this.field_71142_cm = p_i3396_1_.func_71203_ab().func_72395_o();
-        ChunkCoordinates var5 = p_i3396_2_.func_72861_E();
+        ChunkCoordinates var5 = p_i3396_2_.field_73011_w.getRandomizedSpawnPoint();
         int var6 = var5.field_71574_a;
         int var7 = var5.field_71573_c;
         int var8 = var5.field_71572_b;
 
-        if (!p_i3396_2_.field_73011_w.field_76576_e && p_i3396_2_.func_72912_H().func_76077_q() != EnumGameType.ADVENTURE)
-        {
-            int var9 = Math.max(5, p_i3396_1_.func_82357_ak() - 6);
-            var6 += this.field_70146_Z.nextInt(var9 * 2) - var9;
-            var7 += this.field_70146_Z.nextInt(var9 * 2) - var9;
-            var8 = p_i3396_2_.func_72825_h(var6, var7);
-        }
-
         this.func_70012_b((double)var6 + 0.5D, (double)var8, (double)var7 + 0.5D, 0.0F, 0.0F);
         this.field_71133_b = p_i3396_1_;
         this.field_70138_W = 0.0F;
@@ -196,7 +193,10 @@
                 if (var9 != null && this.field_70170_p.func_72899_e(var9.field_77276_a << 4, 0, var9.field_77275_b << 4))
                 {
                     var6.add(this.field_70170_p.func_72964_e(var9.field_77276_a, var9.field_77275_b));
-                    var8.addAll(((WorldServer)this.field_70170_p).func_73049_a(var9.field_77276_a * 16, 0, var9.field_77275_b * 16, var9.field_77276_a * 16 + 16, 256, var9.field_77275_b * 16 + 16));
+                    //BugFix: 16 makes it load an extra chunk, which isn't associated with a player, which makes it not unload unless a player walks near it.
+                    //ToDo: Find a way to efficiently clean abandoned chunks.
+                    //var8.addAll(((WorldServer)this.worldObj).getAllTileEntityInBox(var9.chunkXPos * 16, 0, var9.chunkZPos * 16, var9.chunkXPos * 16 + 16, 256, var9.chunkZPos * 16 + 16));
+                    var8.addAll(((WorldServer)this.field_70170_p).func_73049_a(var9.field_77276_a * 16, 0, var9.field_77275_b * 16, var9.field_77276_a * 16 + 15, 256, var9.field_77275_b * 16 + 15));
                 }
             }
 
@@ -217,6 +217,7 @@
                 {
                     Chunk var11 = (Chunk)var10.next();
                     this.func_71121_q().func_73039_n().func_85172_a(this, var11);
+                    MinecraftForge.EVENT_BUS.post(new ChunkWatchEvent.Watch(var11.func_76632_l(), this));
                 }
             }
         }
@@ -258,11 +259,29 @@
 
     public void func_70645_a(DamageSource p_70645_1_)
     {
+        if (ForgeHooks.onLivingDeath(this, p_70645_1_))
+        {
+            return;
+        }
+        
         this.field_71133_b.func_71203_ab().func_92062_k(p_70645_1_.func_76360_b(this));
 
         if (!this.field_70170_p.func_82736_K().func_82766_b("keepInventory"))
         {
+            captureDrops = true;
+            capturedDrops.clear();
+            
             this.field_71071_by.func_70436_m();
+            
+            captureDrops = false;
+            PlayerDropsEvent event = new PlayerDropsEvent(this, p_70645_1_, capturedDrops, field_70718_bc > 0);
+            if (!MinecraftForge.EVENT_BUS.post(event))
+            {
+                for (EntityItem item : capturedDrops)
+                {
+                    func_71012_a(item);
+                }
+            }
         }
     }
 
