--- a/net/minecraft/entity/player/InventoryPlayer.java
+++ b/net/minecraft/entity/player/InventoryPlayer.java
@@ -287,6 +287,14 @@
                 this.field_70462_a[var1].func_77945_a(this.field_70458_d.field_70170_p, this.field_70458_d, var1, this.field_70461_c == var1);
             }
         }
+        
+        for (int i = 0; i < this.field_70460_b.length; i++)
+        {
+            if (this.field_70460_b[i] != null)
+            {
+                this.field_70460_b[i].func_77973_b().onArmorTickUpdate(this.field_70458_d.field_70170_p, this.field_70458_d, this.field_70460_b[i]);
+            }
+        }
     }
 
     public boolean func_70435_d(int p_70435_1_)
