--- ../src-base/minecraft/net/minecraft/network/NetServerHandler.java
+++ ../src-work/minecraft/net/minecraft/network/NetServerHandler.java
@@ -7,6 +7,8 @@
 import java.util.Iterator;
 import java.util.Random;
 import java.util.logging.Logger;
+
+import cpw.mods.fml.common.network.FMLNetworkHandler;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.item.EntityItem;
 import net.minecraft.entity.player.EntityPlayerMP;
@@ -21,33 +23,7 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.ItemWritableBook;
 import net.minecraft.nbt.NBTTagString;
-import net.minecraft.network.packet.NetHandler;
-import net.minecraft.network.packet.Packet;
-import net.minecraft.network.packet.Packet0KeepAlive;
-import net.minecraft.network.packet.Packet101CloseWindow;
-import net.minecraft.network.packet.Packet102WindowClick;
-import net.minecraft.network.packet.Packet103SetSlot;
-import net.minecraft.network.packet.Packet106Transaction;
-import net.minecraft.network.packet.Packet107CreativeSetSlot;
-import net.minecraft.network.packet.Packet108EnchantItem;
-import net.minecraft.network.packet.Packet10Flying;
-import net.minecraft.network.packet.Packet130UpdateSign;
-import net.minecraft.network.packet.Packet13PlayerLookMove;
-import net.minecraft.network.packet.Packet14BlockDig;
-import net.minecraft.network.packet.Packet15Place;
-import net.minecraft.network.packet.Packet16BlockItemSwitch;
-import net.minecraft.network.packet.Packet18Animation;
-import net.minecraft.network.packet.Packet19EntityAction;
-import net.minecraft.network.packet.Packet202PlayerAbilities;
-import net.minecraft.network.packet.Packet203AutoComplete;
-import net.minecraft.network.packet.Packet204ClientInfo;
-import net.minecraft.network.packet.Packet205ClientCommand;
-import net.minecraft.network.packet.Packet250CustomPayload;
-import net.minecraft.network.packet.Packet255KickDisconnect;
-import net.minecraft.network.packet.Packet3Chat;
-import net.minecraft.network.packet.Packet53BlockChange;
-import net.minecraft.network.packet.Packet7UseEntity;
-import net.minecraft.network.packet.Packet9Respawn;
+import net.minecraft.network.packet.*;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.management.BanEntry;
 import net.minecraft.tileentity.TileEntity;
@@ -60,6 +36,11 @@
 import net.minecraft.util.IntHashMap;
 import net.minecraft.util.MathHelper;
 import net.minecraft.world.WorldServer;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.Event.Result;
+import net.minecraftforge.event.ForgeEventFactory;
+import net.minecraftforge.event.ServerChatEvent;
+import net.minecraftforge.event.entity.player.PlayerInteractEvent;
 
 public class NetServerHandler extends NetHandler
 {
@@ -207,6 +188,11 @@
                     {
                         this.field_72574_e.field_70154_o.func_70043_V();
                     }
+                    
+                    if (!this.field_72587_r) //Fixes teleportation kick while riding entities
+                    {
+                        return;
+                    }
 
                     this.field_72573_d.func_71203_ab().func_72358_d(this.field_72574_e);
                     this.field_72579_o = this.field_72574_e.field_70165_t;
@@ -278,9 +264,10 @@
                 var13 = var5 - this.field_72574_e.field_70165_t;
                 double var15 = var7 - this.field_72574_e.field_70163_u;
                 double var17 = var9 - this.field_72574_e.field_70161_v;
-                double var19 = Math.min(Math.abs(var13), Math.abs(this.field_72574_e.field_70159_w));
-                double var21 = Math.min(Math.abs(var15), Math.abs(this.field_72574_e.field_70181_x));
-                double var23 = Math.min(Math.abs(var17), Math.abs(this.field_72574_e.field_70179_y));
+                //Change min->max, fixes movement speed check 
+                double var19 = Math.max(Math.abs(var13), Math.abs(this.field_72574_e.field_70159_w));
+                double var21 = Math.max(Math.abs(var15), Math.abs(this.field_72574_e.field_70181_x));
+                double var23 = Math.max(Math.abs(var17), Math.abs(this.field_72574_e.field_70179_y));
                 double var25 = var19 * var19 + var21 * var21 + var23 * var23;
 
                 if (var25 > 100.0D && (!this.field_72573_d.func_71264_H() || !this.field_72573_d.func_71214_G().equals(this.field_72574_e.field_71092_bJ)))
@@ -297,6 +284,11 @@
                 {
                     this.field_72574_e.func_71020_j(0.2F);
                 }
+                
+                if (!this.field_72587_r) //Fixes "Moved Too Fast" kick when being teleported while moving
+                {
+                    return;
+                }
 
                 this.field_72574_e.func_70091_d(var13, var15, var17);
                 this.field_72574_e.field_70122_E = p_72498_1_.field_73540_g;
@@ -319,11 +311,16 @@
                     var31 = true;
                     field_72577_a.warning(this.field_72574_e.field_71092_bJ + " moved wrongly!");
                 }
+                
+                if (!this.field_72587_r) //Fixes "Moved Too Fast" kick when being teleported while moving
+                {
+                    return;
+                }
 
                 this.field_72574_e.func_70080_a(var5, var7, var9, var11, var12);
                 boolean var32 = var2.func_72945_a(this.field_72574_e, this.field_72574_e.field_70121_D.func_72329_c().func_72331_e((double)var27, (double)var27, (double)var27)).isEmpty();
 
-                if (var28 && (var31 || !var32) && !this.field_72574_e.func_70608_bn())
+                if (var28 && (var31 || !var32) && !this.field_72574_e.func_70608_bn() && !this.field_72574_e.field_70145_X)
                 {
                     this.func_72569_a(this.field_72579_o, this.field_72589_p, this.field_72588_q, var11, var12);
                     return;
@@ -331,7 +328,7 @@
 
                 AxisAlignedBB var33 = this.field_72574_e.field_70121_D.func_72329_c().func_72314_b((double)var27, (double)var27, (double)var27).func_72321_a(0.0D, -0.55D, 0.0D);
 
-                if (!this.field_72573_d.func_71231_X() && !this.field_72574_e.field_71134_c.func_73083_d() && !var2.func_72829_c(var33))
+                if (!this.field_72573_d.func_71231_X() && !this.field_72574_e.field_71134_c.func_73083_d() && !var2.func_72829_c(var33)  && !this.field_72574_e.field_71075_bZ.field_75101_c)
                 {
                     if (var29 >= -0.03125D)
                     {
@@ -349,6 +346,11 @@
                 {
                     this.field_72572_g = 0;
                 }
+                
+                if (!this.field_72587_r) //Fixes "Moved Too Fast" kick when being teleported while moving
+                {
+                    return;
+                }
 
                 this.field_72574_e.field_70122_E = p_72498_1_.field_73540_g;
                 this.field_72573_d.func_71203_ab().func_72358_d(this.field_72574_e);
@@ -415,7 +417,10 @@
                 double var13 = this.field_72574_e.field_70161_v - ((double)var8 + 0.5D);
                 double var15 = var9 * var9 + var11 * var11 + var13 * var13;
 
-                if (var15 > 36.0D)
+                double dist = field_72574_e.field_71134_c.getBlockReachDistance() + 1;
+                dist *= dist;
+                
+                if (var15 > dist)
                 {
                     return;
                 }
@@ -439,6 +444,7 @@
             {
                 if (var18 <= var3 && !var4)
                 {
+                    ForgeEventFactory.onPlayerInteract(field_72574_e, PlayerInteractEvent.Action.LEFT_CLICK_BLOCK, var6, var7, var8, 0);
                     this.field_72574_e.field_71135_a.func_72567_b(new Packet53BlockChange(var6, var7, var8, var2));
                 }
                 else
@@ -486,7 +492,11 @@
                 return;
             }
 
-            this.field_72574_e.field_71134_c.func_73085_a(this.field_72574_e, var2, var3);
+            PlayerInteractEvent event = ForgeEventFactory.onPlayerInteract(field_72574_e, PlayerInteractEvent.Action.RIGHT_CLICK_AIR, 0, 0, 0, -1);
+            if (event.useItem != Result.DENY)
+            {
+                this.field_72574_e.field_71134_c.func_73085_a(this.field_72574_e, var2, var3);
+            }
         }
         else if (p_72472_1_.func_73402_f() >= this.field_72573_d.func_71207_Z() - 1 && (p_72472_1_.func_73401_h() == 1 || p_72472_1_.func_73402_f() >= this.field_72573_d.func_71207_Z()))
         {
@@ -504,7 +514,9 @@
                 var13 = var12;
             }
 
-            if (this.field_72587_r && this.field_72574_e.func_70092_e((double)var5 + 0.5D, (double)var6 + 0.5D, (double)var7 + 0.5D) < 64.0D && (var13 > var9 || var10))
+            double dist = field_72574_e.field_71134_c.getBlockReachDistance() + 1;
+            dist *= dist;
+            if (this.field_72587_r && this.field_72574_e.func_70092_e((double)var5 + 0.5D, (double)var6 + 0.5D, (double)var7 + 0.5D) < dist && (var13 > var9 || var10))
             {
                 this.field_72574_e.field_71134_c.func_73078_a(this.field_72574_e, var2, var3, var5, var6, var7, var8, p_72472_1_.func_73406_j(), p_72472_1_.func_73404_l(), p_72472_1_.func_73408_m());
             }
@@ -627,6 +639,7 @@
 
     public void func_72481_a(Packet3Chat p_72481_1_)
     {
+        p_72481_1_ = FMLNetworkHandler.handleChatMessage(this, p_72481_1_);
         if (this.field_72574_e.func_71126_v() == 2)
         {
             this.func_72567_b(new Packet3Chat("Cannot send chat message."));
@@ -664,7 +677,12 @@
                         return;
                     }
 
-                    var2 = "<" + this.field_72574_e.field_71092_bJ + "> " + var2;
+                    ServerChatEvent event = new ServerChatEvent(this.field_72574_e, var2, "<" + this.field_72574_e.field_71092_bJ + "> " + var2);
+                    if (MinecraftForge.EVENT_BUS.post(event))
+                    {
+                        return;
+                    }
+                    var2 = event.line;
                     field_72577_a.info(var2);
                     this.field_72573_d.func_71203_ab().func_72384_a(new Packet3Chat(var2, false));
                 }
@@ -786,7 +804,7 @@
                     return;
                 }
 
-                this.field_72574_e = this.field_72573_d.func_71203_ab().func_72368_a(this.field_72574_e, 0, false);
+                this.field_72574_e = this.field_72573_d.func_71203_ab().func_72368_a(this.field_72574_e, field_72574_e.field_71093_bK, false);
             }
         }
     }
@@ -995,6 +1013,11 @@
 
     public void func_72501_a(Packet250CustomPayload p_72501_1_)
     {
+        FMLNetworkHandler.handlePacket250Packet(p_72501_1_, field_72575_b, this);
+    }
+    
+    public void handleVanilla250Packet(Packet250CustomPayload p_72501_1_)
+    {
         DataInputStream var2;
         ItemStack var3;
         ItemStack var4;
@@ -1157,5 +1180,18 @@
                 }
             }
         }
+    }
+    
+    @Override
+    public void func_72494_a(Packet131MapData par1Packet131MapData)
+    {
+        FMLNetworkHandler.handlePacket131Packet(this, par1Packet131MapData);
+    }
+    
+    // modloader compat -- yuk!
+    @Override
+    public EntityPlayerMP getPlayer()
+    {
+        return field_72574_e;
     }
 }
