--- ../src-base/minecraft/net/minecraft/world/chunk/storage/AnvilChunkLoader.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/AnvilChunkLoader.java
@@ -9,6 +9,9 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
+import java.util.logging.Level;
+
+import cpw.mods.fml.common.FMLLog;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityList;
 import net.minecraft.nbt.CompressedStreamTools;
@@ -23,6 +26,8 @@
 import net.minecraft.world.chunk.NibbleArray;
 import net.minecraft.world.storage.IThreadedFileIO;
 import net.minecraft.world.storage.ThreadedFileIOBase;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.world.ChunkDataEvent;
 
 public class AnvilChunkLoader implements IChunkLoader, IThreadedFileIO
 {
@@ -96,6 +101,7 @@
                 var5 = this.func_75823_a(p_75822_1_, p_75822_4_.func_74775_l("Level"));
             }
 
+            MinecraftForge.EVENT_BUS.post(new ChunkDataEvent.Load(var5, p_75822_4_));
             return var5;
         }
     }
@@ -111,6 +117,7 @@
             var3.func_74782_a("Level", var4);
             this.func_75820_a(p_75816_2_, p_75816_1_, var4);
             this.func_75824_a(p_75816_2_.func_76632_l(), var3);
+            MinecraftForge.EVENT_BUS.post(new ChunkDataEvent.Save(p_75816_2_, var3));
         }
         catch (Exception var5)
         {
@@ -253,9 +260,18 @@
                 p_75820_1_.field_76644_m = true;
                 var11 = new NBTTagCompound();
 
-                if (var20.func_70039_c(var11))
-                {
-                    var16.func_74742_a(var11);
+                try
+                {
+                    if (var20.func_70039_c(var11))
+                    {
+                        var16.func_74742_a(var11);
+                    }
+                }
+                catch (Exception e)
+                {
+                    FMLLog.log(Level.SEVERE, e,
+                            "An Entity type %s has thrown an exception trying to write state. It will not persist. Report this to the mod author",
+                            var20.getClass().getName());
                 }
             }
         }
@@ -268,8 +284,17 @@
         {
             TileEntity var21 = (TileEntity)var18.next();
             var11 = new NBTTagCompound();
-            var21.func_70310_b(var11);
-            var17.func_74742_a(var11);
+            try
+            {
+                var21.func_70310_b(var11);
+                var17.func_74742_a(var11);
+            }
+            catch (Exception e)
+            {
+                FMLLog.log(Level.SEVERE, e,
+                        "A TileEntity type %s has throw an exception trying to write state. It will not persist. Report this to the mod author",
+                        var21.getClass().getName());
+            }
         }
 
         p_75820_3_.func_74782_a("TileEntities", var17);
