--- a/net/minecraft/client/MinecraftApplet.java
+++ b/net/minecraft/client/MinecraftApplet.java
@@ -1,5 +1,6 @@
 package net.minecraft.client;
 
+import cpw.mods.fml.relauncher.FMLRelauncher;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.applet.Applet;
@@ -16,6 +17,11 @@
 
     public void init()
     {
+        FMLRelauncher.appletEntry(this);
+    }
+
+    public void fmlInitReentry()
+    {
         this.field_71483_a = new CanvasMinecraftApplet(this);
         boolean var1 = "true".equalsIgnoreCase(this.getParameter("fullscreen"));
         this.field_71481_b = new MinecraftAppletImpl(this, this.field_71483_a, this, this.getWidth(), this.getHeight(), var1);
@@ -61,6 +67,11 @@
     }
 
     public void start()
+    {
+        FMLRelauncher.appletStart(this);
+    }
+
+    public void fmlStartReentry()
     {
         if (this.field_71481_b != null)
         {
