--- a/net/minecraft/item/ItemMap.java
+++ b/net/minecraft/item/ItemMap.java
@@ -52,7 +52,7 @@
             int var5 = 128 * (1 << var4.field_76197_d);
             var4.field_76201_a = Math.round((float)p_77873_2_.func_72912_H().func_76079_c() / (float)var5) * var5;
             var4.field_76199_b = Math.round((float)(p_77873_2_.func_72912_H().func_76074_e() / var5)) * var5;
-            var4.field_76200_c = (byte)p_77873_2_.field_73011_w.field_76574_g;
+            var4.field_76200_c = p_77873_2_.field_73011_w.field_76574_g;
             var4.func_76185_a();
             p_77873_2_.func_72823_a(var3, var4);
         }
@@ -98,7 +98,7 @@
                             boolean var21 = var19 * var19 + var20 * var20 > (var11 - 2) * (var11 - 2);
                             int var22 = (var7 / var6 + var13 - var4 / 2) * var6;
                             int var23 = (var8 / var6 + var18 - var5 / 2) * var6;
-                            int[] var24 = new int[256];
+                            int[] var24 = new int[Block.field_71973_m.length];
                             Chunk var25 = p_77872_1_.func_72938_d(var22, var23);
 
                             if (!var25.func_76621_g())
@@ -194,7 +194,7 @@
                                 var31 = 0;
                                 var32 = 0;
 
-                                for (var33 = 0; var33 < 256; ++var33)
+                                for (var33 = 0; var33 < Block.field_71973_m.length; ++var33)
                                 {
                                     if (var24[var33] > var31)
                                     {
