--- ../src-base/minecraft/net/minecraft/item/ItemBlock.java
+++ ../src-work/minecraft/net/minecraft/item/ItemBlock.java
@@ -18,6 +18,7 @@
         super(p_i3690_1_);
         this.field_77885_a = p_i3690_1_ + 256;
         this.func_77665_c(Block.field_71973_m[p_i3690_1_ + 256].func_71851_a(2));
+        isDefaultTexture = Block.field_71973_m[p_i3690_1_ + 256].isDefaultTexture;
     }
 
     public int func_77883_f()
@@ -33,7 +34,8 @@
         {
             p_77648_7_ = 1;
         }
-        else if (var11 != Block.field_71998_bu.field_71990_ca && var11 != Block.field_71962_X.field_71990_ca && var11 != Block.field_71961_Y.field_71990_ca)
+        else if (var11 != Block.field_71998_bu.field_71990_ca && var11 != Block.field_71962_X.field_71990_ca && var11 != Block.field_71961_Y.field_71990_ca
+                && (Block.field_71973_m[var11] == null || !Block.field_71973_m[var11].isBlockReplaceable(p_77648_3_, p_77648_4_, p_77648_5_, p_77648_6_)))
         {
             if (p_77648_7_ == 0)
             {
@@ -84,14 +86,8 @@
             int var13 = this.func_77647_b(p_77648_1_.func_77960_j());
             int var14 = Block.field_71973_m[this.field_77885_a].func_85104_a(p_77648_3_, p_77648_4_, p_77648_5_, p_77648_6_, p_77648_7_, p_77648_8_, p_77648_9_, p_77648_10_, var13);
 
-            if (p_77648_3_.func_72832_d(p_77648_4_, p_77648_5_, p_77648_6_, this.field_77885_a, var14))
+            if (placeBlockAt(p_77648_1_, p_77648_2_, p_77648_3_, p_77648_4_, p_77648_5_, p_77648_6_, p_77648_7_, p_77648_8_, p_77648_9_, p_77648_10_, var14))
             {
-                if (p_77648_3_.func_72798_a(p_77648_4_, p_77648_5_, p_77648_6_) == this.field_77885_a)
-                {
-                    Block.field_71973_m[this.field_77885_a].func_71860_a(p_77648_3_, p_77648_4_, p_77648_5_, p_77648_6_, p_77648_2_);
-                    Block.field_71973_m[this.field_77885_a].func_85105_g(p_77648_3_, p_77648_4_, p_77648_5_, p_77648_6_, var14);
-                }
-
                 p_77648_3_.func_72908_a((double)((float)p_77648_4_ + 0.5F), (double)((float)p_77648_5_ + 0.5F), (double)((float)p_77648_6_ + 0.5F), var12.field_72020_cn.func_82593_b(), (var12.field_72020_cn.func_72677_b() + 1.0F) / 2.0F, var12.field_72020_cn.func_72678_c() * 0.8F);
                 --p_77648_1_.field_77994_a;
             }
@@ -113,7 +109,8 @@
         {
             p_77884_5_ = 1;
         }
-        else if (var8 != Block.field_71998_bu.field_71990_ca && var8 != Block.field_71962_X.field_71990_ca && var8 != Block.field_71961_Y.field_71990_ca)
+        else if (var8 != Block.field_71998_bu.field_71990_ca && var8 != Block.field_71962_X.field_71990_ca && var8 != Block.field_71961_Y.field_71990_ca
+                && (Block.field_71973_m[var8] == null || !Block.field_71973_m[var8].isBlockReplaceable(p_77884_1_, p_77884_2_, p_77884_3_, p_77884_4_)))
         {
             if (p_77884_5_ == 0)
             {
@@ -169,5 +166,29 @@
     public void func_77633_a(int p_77633_1_, CreativeTabs p_77633_2_, List p_77633_3_)
     {
         Block.field_71973_m[this.field_77885_a].func_71879_a(p_77633_1_, p_77633_2_, p_77633_3_);
+    }
+    
+    /**
+     * Called to actually place the block, after the location is determined
+     * and all permission checks have been made.
+     *
+     * @param stack The item stack that was used to place the block. This can be changed inside the method.
+     * @param player The player who is placing the block. Can be null if the block is not being placed by a player.
+     * @param side The side the player (or machine) right-clicked on.
+     */
+    public boolean placeBlockAt(ItemStack stack, EntityPlayer player, World world, int x, int y, int z, int side, float hitX, float hitY, float hitZ, int metadata)
+    {
+        if (!world.func_72832_d(x, y, z, this.field_77885_a, metadata))
+        {
+            return false;
+        }
+    
+        if (world.func_72798_a(x, y, z) == this.field_77885_a)
+        {
+            Block.field_71973_m[this.field_77885_a].func_71860_a(world, x, y, z, player);
+            Block.field_71973_m[this.field_77885_a].func_85105_g(world, x, y, z, metadata);
+        }
+    
+        return true;
     }
 }
