--- ../src-base/minecraft/net/minecraft/world/gen/structure/ComponentScatteredFeatureJunglePyramid.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/ComponentScatteredFeatureJunglePyramid.java
@@ -7,6 +7,9 @@
 import net.minecraft.util.WeightedRandomChestContent;
 import net.minecraft.world.World;
 
+import net.minecraftforge.common.ChestGenHooks;
+import static net.minecraftforge.common.ChestGenHooks.*;
+
 public class ComponentScatteredFeatureJunglePyramid extends ComponentScatteredFeature
 {
     private boolean field_74947_h;
@@ -152,9 +155,12 @@
             this.func_74864_a(p_74875_1_, Block.field_72075_av.field_71990_ca, 0, 4, -3, 1, p_74875_3_);
             this.func_74864_a(p_74875_1_, Block.field_72087_ao.field_71990_ca, 0, 3, -3, 1, p_74875_3_);
 
+            ChestGenHooks dispenser = ChestGenHooks.getInfo(PYRAMID_JUNGLE_DISPENSER);
+            ChestGenHooks chest = ChestGenHooks.getInfo(PYRAMID_JUNGLE_CHEST);
+            
             if (!this.field_74945_j)
             {
-                this.field_74945_j = this.func_74869_a(p_74875_1_, p_74875_3_, p_74875_2_, 3, -2, 1, 2, field_74944_m, 2);
+                this.field_74945_j = this.func_74869_a(p_74875_1_, p_74875_3_, p_74875_2_, 3, -2, 1, 2, dispenser.getItems(p_74875_2_), dispenser.getCount(p_74875_2_));
             }
 
             this.func_74864_a(p_74875_1_, Block.field_71998_bu.field_71990_ca, 15, 3, -2, 2, p_74875_3_);
@@ -171,7 +177,7 @@
 
             if (!this.field_74946_k)
             {
-                this.field_74946_k = this.func_74869_a(p_74875_1_, p_74875_3_, p_74875_2_, 9, -2, 3, 4, field_74944_m, 2);
+                this.field_74946_k = this.func_74869_a(p_74875_1_, p_74875_3_, p_74875_2_, 9, -2, 3, 4, dispenser.getItems(p_74875_2_), dispenser.getCount(p_74875_2_));
             }
 
             this.func_74864_a(p_74875_1_, Block.field_71998_bu.field_71990_ca, 15, 8, -1, 3, p_74875_3_);
@@ -179,7 +185,7 @@
 
             if (!this.field_74947_h)
             {
-                this.field_74947_h = this.func_74879_a(p_74875_1_, p_74875_3_, p_74875_2_, 8, -3, 3, WeightedRandomChestContent.func_92080_a(field_74943_l, new WeightedRandomChestContent[] {Item.field_92105_bW.func_92114_b(p_74875_2_)}), 2 + p_74875_2_.nextInt(5));
+                this.field_74947_h = this.func_74879_a(p_74875_1_, p_74875_3_, p_74875_2_, 8, -3, 3, chest.getItems(p_74875_2_), chest.getCount(p_74875_2_));;
             }
 
             this.func_74864_a(p_74875_1_, Block.field_72087_ao.field_71990_ca, 0, 9, -3, 2, p_74875_3_);
@@ -212,7 +218,7 @@
 
             if (!this.field_74948_i)
             {
-                this.field_74948_i = this.func_74879_a(p_74875_1_, p_74875_3_, p_74875_2_, 9, -3, 10, WeightedRandomChestContent.func_92080_a(field_74943_l, new WeightedRandomChestContent[] {Item.field_92105_bW.func_92114_b(p_74875_2_)}), 2 + p_74875_2_.nextInt(5));
+                this.field_74948_i = this.func_74879_a(p_74875_1_, p_74875_3_, p_74875_2_, 9, -3, 10, chest.getItems(p_74875_2_), chest.getCount(p_74875_2_));
             }
 
             return true;
