--- a/net/minecraft/stats/StatList.java
+++ b/net/minecraft/stats/StatList.java
@@ -118,9 +118,9 @@
 
     private static StatBase[] func_75921_a(String p_75921_0_, int p_75921_1_)
     {
-        StatBase[] var2 = new StatBase[256];
-
-        for (int var3 = 0; var3 < 256; ++var3)
+        StatBase[] var2 = new StatBase[Block.field_71973_m.length];
+        
+        for (int var3 = 0; var3 < Block.field_71973_m.length; ++var3)
         {
             if (Block.field_71973_m[var3] != null && Block.field_71973_m[var3].func_71876_u())
             {
