--- ../src-base/minecraft/net/minecraft/enchantment/Enchantment.java
+++ ../src-work/minecraft/net/minecraft/enchantment/Enchantment.java
@@ -1,6 +1,8 @@
 package net.minecraft.enchantment;
 
 import java.util.ArrayList;
+
+import com.google.common.collect.ObjectArrays;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.DamageSource;
@@ -113,6 +115,27 @@
     public boolean func_92089_a(ItemStack p_92089_1_)
     {
         return this.field_77351_y.func_77557_a(p_92089_1_.func_77973_b());
+    }
+    
+    /**
+     * This applies specifically to applying at the enchanting table. The other method {@link #func_92037_a(ItemStack)}
+     * applies for <i>all possible</i> enchantments.
+     * @param stack
+     * @return
+     */
+    public boolean canApplyAtEnchantingTable(ItemStack stack)
+    {
+        return this.field_77351_y.func_77557_a(stack.func_77973_b());
+    }
+    
+    /**
+     * Add to the list of enchantments applicable by the anvil from a book
+     *
+     * @param enchantment
+     */
+    public static void addToBookList(Enchantment enchantment)
+    {
+        ObjectArrays.concat(field_92090_c, enchantment);
     }
 
     static
