--- a/net/minecraft/block/BlockSkull.java
+++ b/net/minecraft/block/BlockSkull.java
@@ -2,6 +2,8 @@
 
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
+
+import java.util.ArrayList;
 import java.util.Random;
 import net.minecraft.block.material.Material;
 import net.minecraft.entity.EntityLiving;
@@ -99,10 +101,6 @@
         return p_71899_1_;
     }
 
-    public void func_71914_a(World p_71914_1_, int p_71914_2_, int p_71914_3_, int p_71914_4_, int p_71914_5_, float p_71914_6_, int p_71914_7_)
-    {
-    }
-
     public void func_71846_a(World p_71846_1_, int p_71846_2_, int p_71846_3_, int p_71846_4_, int p_71846_5_, EntityPlayer p_71846_6_)
     {
         if (p_71846_6_.field_71075_bZ.field_75098_d)
@@ -110,30 +108,38 @@
             p_71846_5_ |= 8;
             p_71846_1_.func_72921_c(p_71846_2_, p_71846_3_, p_71846_4_, p_71846_5_);
         }
+        
+        func_71897_c(p_71846_1_, p_71846_2_, p_71846_3_, p_71846_4_, p_71846_5_, 0);
 
         super.func_71846_a(p_71846_1_, p_71846_2_, p_71846_3_, p_71846_4_, p_71846_5_, p_71846_6_);
     }
 
     public void func_71852_a(World p_71852_1_, int p_71852_2_, int p_71852_3_, int p_71852_4_, int p_71852_5_, int p_71852_6_)
     {
-        if (!p_71852_1_.field_72995_K)
+        super.func_71852_a(p_71852_1_, p_71852_2_, p_71852_3_, p_71852_4_, p_71852_5_, p_71852_6_);
+    }
+    
+    @Override
+    public ArrayList<ItemStack> getBlockDropped(World world, int x, int y, int z, int metadata, int fortune)
+    {
+        ArrayList<ItemStack> drops = new ArrayList<ItemStack>();
+        if ((metadata & 8) == 0)
         {
-            if ((p_71852_6_ & 8) == 0)
-            {
-                ItemStack var7 = new ItemStack(Item.field_82799_bQ.field_77779_bT, 1, this.func_71873_h(p_71852_1_, p_71852_2_, p_71852_3_, p_71852_4_));
-                TileEntitySkull var8 = (TileEntitySkull)p_71852_1_.func_72796_p(p_71852_2_, p_71852_3_, p_71852_4_);
-
-                if (var8.func_82117_a() == 3 && var8.func_82120_c() != null && var8.func_82120_c().length() > 0)
-                {
-                    var7.func_77982_d(new NBTTagCompound());
-                    var7.func_77978_p().func_74778_a("SkullOwner", var8.func_82120_c());
-                }
-
-                this.func_71929_a(p_71852_1_, p_71852_2_, p_71852_3_, p_71852_4_, var7);
-            }
-
-            super.func_71852_a(p_71852_1_, p_71852_2_, p_71852_3_, p_71852_4_, p_71852_5_, p_71852_6_);
+            ItemStack var7 = new ItemStack(Item.field_82799_bQ.field_77779_bT, 1, this.func_71873_h(world, x, y, z));
+            TileEntitySkull var8 = (TileEntitySkull)world.func_72796_p(x, y, z);
+        
+            if (var8 == null)
+            {
+                return drops;
+            }
+            if (var8.func_82117_a() == 3 && var8.func_82120_c() != null && var8.func_82120_c().length() > 0)
+            {
+                var7.func_77982_d(new NBTTagCompound());
+                var7.func_77978_p().func_74778_a("SkullOwner", var8.func_82120_c());
+            }
+            drops.add(var7);
         }
+        return drops;
     }
 
     public int func_71885_a(int p_71885_1_, Random p_71885_2_, int p_71885_3_)
