--- a/net/minecraft/block/BlockLever.java
+++ b/net/minecraft/block/BlockLever.java
@@ -7,6 +7,9 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import net.minecraftforge.common.ForgeDirection;
+import static net.minecraftforge.common.ForgeDirection.*;
+
 public class BlockLever extends Block
 {
     protected BlockLever(int p_i3962_1_, int p_i3962_2_)
@@ -37,12 +40,23 @@
 
     public boolean func_71850_a_(World p_71850_1_, int p_71850_2_, int p_71850_3_, int p_71850_4_, int p_71850_5_)
     {
-        return p_71850_5_ == 0 && p_71850_1_.func_72809_s(p_71850_2_, p_71850_3_ + 1, p_71850_4_) ? true : (p_71850_5_ == 1 && p_71850_1_.func_72797_t(p_71850_2_, p_71850_3_ - 1, p_71850_4_) ? true : (p_71850_5_ == 2 && p_71850_1_.func_72809_s(p_71850_2_, p_71850_3_, p_71850_4_ + 1) ? true : (p_71850_5_ == 3 && p_71850_1_.func_72809_s(p_71850_2_, p_71850_3_, p_71850_4_ - 1) ? true : (p_71850_5_ == 4 && p_71850_1_.func_72809_s(p_71850_2_ + 1, p_71850_3_, p_71850_4_) ? true : p_71850_5_ == 5 && p_71850_1_.func_72809_s(p_71850_2_ - 1, p_71850_3_, p_71850_4_)))));
+        ForgeDirection dir = ForgeDirection.getOrientation(p_71850_5_);
+        return (dir == DOWN  && p_71850_1_.isBlockSolidOnSide(p_71850_2_, p_71850_3_ + 1, p_71850_4_, DOWN )) ||
+                (dir == UP    && p_71850_1_.isBlockSolidOnSide(p_71850_2_, p_71850_3_ - 1, p_71850_4_, UP   )) ||
+                (dir == NORTH && p_71850_1_.isBlockSolidOnSide(p_71850_2_, p_71850_3_, p_71850_4_ + 1, NORTH)) ||
+                (dir == SOUTH && p_71850_1_.isBlockSolidOnSide(p_71850_2_, p_71850_3_, p_71850_4_ - 1, SOUTH)) ||
+                (dir == WEST  && p_71850_1_.isBlockSolidOnSide(p_71850_2_ + 1, p_71850_3_, p_71850_4_, WEST )) ||
+                (dir == EAST  && p_71850_1_.isBlockSolidOnSide(p_71850_2_ - 1, p_71850_3_, p_71850_4_, EAST ));
     }
 
     public boolean func_71930_b(World p_71930_1_, int p_71930_2_, int p_71930_3_, int p_71930_4_)
     {
-        return p_71930_1_.func_72809_s(p_71930_2_ - 1, p_71930_3_, p_71930_4_) ? true : (p_71930_1_.func_72809_s(p_71930_2_ + 1, p_71930_3_, p_71930_4_) ? true : (p_71930_1_.func_72809_s(p_71930_2_, p_71930_3_, p_71930_4_ - 1) ? true : (p_71930_1_.func_72809_s(p_71930_2_, p_71930_3_, p_71930_4_ + 1) ? true : (p_71930_1_.func_72797_t(p_71930_2_, p_71930_3_ - 1, p_71930_4_) ? true : p_71930_1_.func_72809_s(p_71930_2_, p_71930_3_ + 1, p_71930_4_)))));
+        return p_71930_1_.isBlockSolidOnSide(p_71930_2_ - 1, p_71930_3_, p_71930_4_, EAST ) ||
+                p_71930_1_.isBlockSolidOnSide(p_71930_2_ + 1, p_71930_3_, p_71930_4_, WEST ) ||
+                p_71930_1_.isBlockSolidOnSide(p_71930_2_, p_71930_3_, p_71930_4_ - 1, SOUTH) ||
+                p_71930_1_.isBlockSolidOnSide(p_71930_2_, p_71930_3_, p_71930_4_ + 1, NORTH) ||
+                p_71930_1_.isBlockSolidOnSide(p_71930_2_, p_71930_3_ - 1, p_71930_4_, UP   ) ||
+                p_71930_1_.isBlockSolidOnSide(p_71930_2_, p_71930_3_ + 1, p_71930_4_, DOWN );
     }
 
     public int func_85104_a(World p_85104_1_, int p_85104_2_, int p_85104_3_, int p_85104_4_, int p_85104_5_, float p_85104_6_, float p_85104_7_, float p_85104_8_, int p_85104_9_)
@@ -51,32 +65,32 @@
         int var10 = p_85104_9_ & 7;
         var10 = -1;
 
-        if (p_85104_5_ == 0 && p_85104_1_.func_72809_s(p_85104_2_, p_85104_3_ + 1, p_85104_4_))
+        if (p_85104_5_ == 0 && p_85104_1_.isBlockSolidOnSide(p_85104_2_, p_85104_3_ + 1, p_85104_4_, DOWN))
         {
             var10 = p_85104_1_.field_73012_v.nextBoolean() ? 0 : 7;
         }
 
-        if (p_85104_5_ == 1 && p_85104_1_.func_72797_t(p_85104_2_, p_85104_3_ - 1, p_85104_4_))
+        if (p_85104_5_ == 1 && p_85104_1_.isBlockSolidOnSide(p_85104_2_, p_85104_3_ - 1, p_85104_4_, UP))
         {
             var10 = 5 + p_85104_1_.field_73012_v.nextInt(2);
         }
 
-        if (p_85104_5_ == 2 && p_85104_1_.func_72809_s(p_85104_2_, p_85104_3_, p_85104_4_ + 1))
+        if (p_85104_5_ == 2 && p_85104_1_.isBlockSolidOnSide(p_85104_2_, p_85104_3_, p_85104_4_ + 1, NORTH))
         {
             var10 = 4;
         }
 
-        if (p_85104_5_ == 3 && p_85104_1_.func_72809_s(p_85104_2_, p_85104_3_, p_85104_4_ - 1))
+        if (p_85104_5_ == 3 && p_85104_1_.isBlockSolidOnSide(p_85104_2_, p_85104_3_, p_85104_4_ - 1, SOUTH))
         {
             var10 = 3;
         }
 
-        if (p_85104_5_ == 4 && p_85104_1_.func_72809_s(p_85104_2_ + 1, p_85104_3_, p_85104_4_))
+        if (p_85104_5_ == 4 && p_85104_1_.isBlockSolidOnSide(p_85104_2_ + 1, p_85104_3_, p_85104_4_, WEST))
         {
             var10 = 2;
         }
 
-        if (p_85104_5_ == 5 && p_85104_1_.func_72809_s(p_85104_2_ - 1, p_85104_3_, p_85104_4_))
+        if (p_85104_5_ == 5 && p_85104_1_.isBlockSolidOnSide(p_85104_2_ - 1, p_85104_3_, p_85104_4_, EAST))
         {
             var10 = 1;
         }
@@ -112,42 +126,42 @@
             int var6 = p_71863_1_.func_72805_g(p_71863_2_, p_71863_3_, p_71863_4_) & 7;
             boolean var7 = false;
 
-            if (!p_71863_1_.func_72809_s(p_71863_2_ - 1, p_71863_3_, p_71863_4_) && var6 == 1)
-            {
-                var7 = true;
-            }
-
-            if (!p_71863_1_.func_72809_s(p_71863_2_ + 1, p_71863_3_, p_71863_4_) && var6 == 2)
-            {
-                var7 = true;
-            }
-
-            if (!p_71863_1_.func_72809_s(p_71863_2_, p_71863_3_, p_71863_4_ - 1) && var6 == 3)
-            {
-                var7 = true;
-            }
-
-            if (!p_71863_1_.func_72809_s(p_71863_2_, p_71863_3_, p_71863_4_ + 1) && var6 == 4)
-            {
-                var7 = true;
-            }
-
-            if (!p_71863_1_.func_72797_t(p_71863_2_, p_71863_3_ - 1, p_71863_4_) && var6 == 5)
-            {
-                var7 = true;
-            }
-
-            if (!p_71863_1_.func_72797_t(p_71863_2_, p_71863_3_ - 1, p_71863_4_) && var6 == 6)
-            {
-                var7 = true;
-            }
-
-            if (!p_71863_1_.func_72809_s(p_71863_2_, p_71863_3_ + 1, p_71863_4_) && var6 == 0)
-            {
-                var7 = true;
-            }
-
-            if (!p_71863_1_.func_72809_s(p_71863_2_, p_71863_3_ + 1, p_71863_4_) && var6 == 7)
+            if (!p_71863_1_.isBlockSolidOnSide(p_71863_2_ - 1, p_71863_3_, p_71863_4_, EAST) && var6 == 1)
+            {
+                var7 = true;
+            }
+            
+            if (!p_71863_1_.isBlockSolidOnSide(p_71863_2_ + 1, p_71863_3_, p_71863_4_, WEST) && var6 == 2)
+            {
+                var7 = true;
+            }
+            
+            if (!p_71863_1_.isBlockSolidOnSide(p_71863_2_, p_71863_3_, p_71863_4_ - 1, SOUTH) && var6 == 3)
+            {
+                var7 = true;
+            }
+            
+            if (!p_71863_1_.isBlockSolidOnSide(p_71863_2_, p_71863_3_, p_71863_4_ + 1, NORTH) && var6 == 4)
+            {
+                var7 = true;
+            }
+            
+            if (!p_71863_1_.isBlockSolidOnSide(p_71863_2_, p_71863_3_ - 1, p_71863_4_, UP) && var6 == 5)
+            {
+                var7 = true;
+            }
+            
+            if (!p_71863_1_.isBlockSolidOnSide(p_71863_2_, p_71863_3_ - 1, p_71863_4_, UP) && var6 == 6)
+            {
+                var7 = true;
+            }
+            
+            if (!p_71863_1_.isBlockSolidOnSide(p_71863_2_, p_71863_3_ + 1, p_71863_4_, DOWN) && var6 == 0)
+            {
+                var7 = true;
+            }
+            
+            if (!p_71863_1_.isBlockSolidOnSide(p_71863_2_, p_71863_3_ + 1, p_71863_4_, DOWN) && var6 == 7)
             {
                 var7 = true;
             }
