--- a/net/minecraft/block/BlockChest.java
+++ b/net/minecraft/block/BlockChest.java
@@ -21,6 +21,8 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import static net.minecraftforge.common.ForgeDirection.*;
+
 public class BlockChest extends BlockContainer
 {
     private Random field_72293_a = new Random();
@@ -382,7 +384,7 @@
         {
             return true;
         }
-        else if (p_71903_1_.func_72809_s(p_71903_2_, p_71903_3_ + 1, p_71903_4_))
+        else if (p_71903_1_.isBlockSolidOnSide(p_71903_2_, p_71903_3_ + 1, p_71903_4_, DOWN))
         {
             return true;
         }
@@ -390,19 +392,19 @@
         {
             return true;
         }
-        else if (p_71903_1_.func_72798_a(p_71903_2_ - 1, p_71903_3_, p_71903_4_) == this.field_71990_ca && (p_71903_1_.func_72809_s(p_71903_2_ - 1, p_71903_3_ + 1, p_71903_4_) || func_72292_n(p_71903_1_, p_71903_2_ - 1, p_71903_3_, p_71903_4_)))
-        {
-            return true;
-        }
-        else if (p_71903_1_.func_72798_a(p_71903_2_ + 1, p_71903_3_, p_71903_4_) == this.field_71990_ca && (p_71903_1_.func_72809_s(p_71903_2_ + 1, p_71903_3_ + 1, p_71903_4_) || func_72292_n(p_71903_1_, p_71903_2_ + 1, p_71903_3_, p_71903_4_)))
-        {
-            return true;
-        }
-        else if (p_71903_1_.func_72798_a(p_71903_2_, p_71903_3_, p_71903_4_ - 1) == this.field_71990_ca && (p_71903_1_.func_72809_s(p_71903_2_, p_71903_3_ + 1, p_71903_4_ - 1) || func_72292_n(p_71903_1_, p_71903_2_, p_71903_3_, p_71903_4_ - 1)))
-        {
-            return true;
-        }
-        else if (p_71903_1_.func_72798_a(p_71903_2_, p_71903_3_, p_71903_4_ + 1) == this.field_71990_ca && (p_71903_1_.func_72809_s(p_71903_2_, p_71903_3_ + 1, p_71903_4_ + 1) || func_72292_n(p_71903_1_, p_71903_2_, p_71903_3_, p_71903_4_ + 1)))
+        else if (p_71903_1_.func_72798_a(p_71903_2_ - 1, p_71903_3_, p_71903_4_) == this.field_71990_ca && (p_71903_1_.isBlockSolidOnSide(p_71903_2_ - 1, p_71903_3_ + 1, p_71903_4_, DOWN) || func_72292_n(p_71903_1_, p_71903_2_ - 1, p_71903_3_, p_71903_4_)))
+        {
+            return true;
+        }
+        else if (p_71903_1_.func_72798_a(p_71903_2_ + 1, p_71903_3_, p_71903_4_) == this.field_71990_ca && (p_71903_1_.isBlockSolidOnSide(p_71903_2_ + 1, p_71903_3_ + 1, p_71903_4_, DOWN) || func_72292_n(p_71903_1_, p_71903_2_ + 1, p_71903_3_, p_71903_4_)))
+        {
+            return true;
+        }
+        else if (p_71903_1_.func_72798_a(p_71903_2_, p_71903_3_, p_71903_4_ - 1) == this.field_71990_ca && (p_71903_1_.isBlockSolidOnSide(p_71903_2_, p_71903_3_ + 1, p_71903_4_ - 1, DOWN) || func_72292_n(p_71903_1_, p_71903_2_, p_71903_3_, p_71903_4_ - 1)))
+        {
+            return true;
+        }
+        else if (p_71903_1_.func_72798_a(p_71903_2_, p_71903_3_, p_71903_4_ + 1) == this.field_71990_ca && (p_71903_1_.isBlockSolidOnSide(p_71903_2_, p_71903_3_ + 1, p_71903_4_ + 1, DOWN) || func_72292_n(p_71903_1_, p_71903_2_, p_71903_3_, p_71903_4_ + 1)))
         {
             return true;
         }
