--- a/net/minecraft/world/storage/MapData.java
+++ b/net/minecraft/world/storage/MapData.java
@@ -9,7 +9,7 @@
 import java.util.Map;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.ItemStack;
-import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.*;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldSavedData;
 
@@ -17,7 +17,7 @@
 {
     public int field_76201_a;
     public int field_76199_b;
-    public byte field_76200_c;
+    public int field_76200_c;
     public byte field_76197_d;
     public byte[] field_76198_e = new byte[16384];
     public List field_76196_g = new ArrayList();
@@ -31,7 +31,17 @@
 
     public void func_76184_a(NBTTagCompound p_76184_1_)
     {
-        this.field_76200_c = p_76184_1_.func_74771_c("dimension");
+        NBTBase dimension = p_76184_1_.func_74781_a("dimension");
+
+        if (dimension instanceof NBTTagByte)
+        {
+            this.field_76200_c = ((NBTTagByte)dimension).field_74756_a;
+        }
+        else
+        {
+            this.field_76200_c = ((NBTTagInt)dimension).field_74748_a;
+        }
+
         this.field_76201_a = p_76184_1_.func_74762_e("xCenter");
         this.field_76199_b = p_76184_1_.func_74762_e("zCenter");
         this.field_76197_d = p_76184_1_.func_74771_c("scale");
@@ -82,7 +92,7 @@
 
     public void func_76187_b(NBTTagCompound p_76187_1_)
     {
-        p_76187_1_.func_74774_a("dimension", this.field_76200_c);
+        p_76187_1_.func_74768_a("dimension", this.field_76200_c);
         p_76187_1_.func_74768_a("xCenter", this.field_76201_a);
         p_76187_1_.func_74768_a("zCenter", this.field_76199_b);
         p_76187_1_.func_74774_a("scale", this.field_76197_d);
@@ -144,7 +154,7 @@
             p_82567_8_ += p_82567_8_ < 0.0D ? -8.0D : 8.0D;
             var15 = (byte)((int)(p_82567_8_ * 16.0D / 360.0D));
 
-            if (this.field_76200_c < 0)
+            if (p_82567_2_.field_73011_w.shouldMapSpin(p_82567_3_, p_82567_4_, p_82567_6_, p_82567_8_))
             {
                 int var17 = (int)(p_82567_2_.func_72912_H().func_76073_f() / 10L);
                 var15 = (byte)(var17 * var17 * 34187121 + var17 * 121 >> 15 & 15);
