--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenShrub.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenShrub.java
@@ -19,10 +19,16 @@
     {
         int var15;
 
-        for (boolean var6 = false; ((var15 = p_76484_1_.func_72798_a(p_76484_3_, p_76484_4_, p_76484_5_)) == 0 || var15 == Block.field_71952_K.field_71990_ca) && p_76484_4_ > 0; --p_76484_4_)
+        Block block = null;
+        do 
         {
-            ;
-        }
+            block = Block.field_71973_m[p_76484_1_.func_72798_a(p_76484_3_,  p_76484_4_, p_76484_5_)];
+            if (block != null && !block.isLeaves(p_76484_1_, p_76484_3_, p_76484_4_, p_76484_5_))
+            {
+                break;
+            }
+            p_76484_4_--;
+        } while (p_76484_4_ > 0);
 
         int var7 = p_76484_1_.func_72798_a(p_76484_3_, p_76484_4_, p_76484_5_);
 
@@ -44,7 +50,10 @@
                     {
                         int var14 = var13 - p_76484_5_;
 
-                        if ((Math.abs(var12) != var10 || Math.abs(var14) != var10 || p_76484_2_.nextInt(2) != 0) && !Block.field_71970_n[p_76484_1_.func_72798_a(var11, var8, var13)])
+                        block = Block.field_71973_m[p_76484_1_.func_72798_a(var11, var8, var13)];
+                        
+                        if ((Math.abs(var12) != var10 || Math.abs(var14) != var10 || p_76484_2_.nextInt(2) != 0) && 
+                                (block == null || block.canBeReplacedByLeaves(p_76484_1_, var11, var8, var13)))
                         {
                             this.func_76485_a(p_76484_1_, var11, var8, var13, Block.field_71952_K.field_71990_ca, this.field_76527_a);
                         }
