--- ../src-base/minecraft/net/minecraft/client/renderer/entity/RenderBiped.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/RenderBiped.java
@@ -9,11 +9,13 @@
 import net.minecraft.client.renderer.tileentity.TileEntitySkullRenderer;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLiving;
-import net.minecraft.item.EnumArmorMaterial;
-import net.minecraft.item.Item;
-import net.minecraft.item.ItemArmor;
-import net.minecraft.item.ItemStack;
+import net.minecraft.item.*;
 import org.lwjgl.opengl.GL11;
+import static net.minecraftforge.client.IItemRenderer.ItemRenderType.EQUIPPED;
+import static net.minecraftforge.client.IItemRenderer.ItemRendererHelper.BLOCK_3D;
+import net.minecraftforge.client.ForgeHooksClient;
+import net.minecraftforge.client.IItemRenderer;
+import net.minecraftforge.client.MinecraftForgeClient;
 
 @SideOnly(Side.CLIENT)
 public class RenderBiped extends RenderLiving
@@ -54,7 +56,7 @@
             if (var5 instanceof ItemArmor)
             {
                 ItemArmor var6 = (ItemArmor)var5;
-                this.func_76985_a("/armor/" + field_82424_k[var6.field_77880_c] + "_" + (p_77032_2_ == 2 ? 2 : 1) + ".png");
+                this.func_76985_a(ForgeHooksClient.getArmorTexture(var4, "/armor/" + field_82424_k[var6.field_77880_c] + "_" + (p_77032_2_ == 2 ? 2 : 1) + ".png"));
                 ModelBiped var7 = p_77032_2_ == 2 ? this.field_82425_h : this.field_82423_g;
                 var7.field_78116_c.field_78806_j = p_77032_2_ == 0;
                 var7.field_78114_d.field_78806_j = p_77032_2_ == 0;
@@ -169,9 +171,12 @@
             GL11.glPushMatrix();
             this.field_77071_a.field_78116_c.func_78794_c(0.0625F);
 
-            if (var5.func_77973_b().field_77779_bT < 256)
+            IItemRenderer customRenderer = MinecraftForgeClient.getItemRenderer(var5, EQUIPPED);
+            boolean is3D = (customRenderer != null && customRenderer.shouldUseRenderHelper(EQUIPPED, var5, BLOCK_3D));
+            
+            if (var5.func_77973_b() instanceof ItemBlock)
             {
-                if (RenderBlocks.func_78597_b(Block.field_71973_m[var5.field_77993_c].func_71857_b()))
+                if (is3D || RenderBlocks.func_78597_b(Block.field_71973_m[var5.field_77993_c].func_71857_b()))
                 {
                     var6 = 0.625F;
                     GL11.glTranslatef(0.0F, -0.25F, 0.0F);
@@ -213,7 +218,10 @@
             this.field_77071_a.field_78112_f.func_78794_c(0.0625F);
             GL11.glTranslatef(-0.0625F, 0.4375F, 0.0625F);
 
-            if (var4.field_77993_c < 256 && RenderBlocks.func_78597_b(Block.field_71973_m[var4.field_77993_c].func_71857_b()))
+            IItemRenderer customRenderer = MinecraftForgeClient.getItemRenderer(var4, EQUIPPED);
+            boolean is3D = (customRenderer != null && customRenderer.shouldUseRenderHelper(EQUIPPED, var4, BLOCK_3D));
+            
+            if (var4.func_77973_b() instanceof ItemBlock && (is3D || RenderBlocks.func_78597_b(Block.field_71973_m[var4.field_77993_c].func_71857_b())))
             {
                 var6 = 0.5F;
                 GL11.glTranslatef(0.0F, 0.1875F, -0.3125F);
@@ -260,7 +268,10 @@
 
             if (var4.func_77973_b().func_77623_v())
             {
-                this.field_76990_c.field_78721_f.func_78443_a(p_77029_1_, var4, 1);
+                for (int x = 1; x < var4.func_77973_b().getRenderPasses(var4.func_77960_j()); x++)
+                {
+                    this.field_76990_c.field_78721_f.func_78443_a(p_77029_1_, var4, x);
+                }
             }
 
             GL11.glPopMatrix();
