--- ../src-base/minecraft/net/minecraft/client/renderer/entity/RenderManager.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/entity/RenderManager.java
@@ -198,12 +198,14 @@
 
         if (p_78718_4_.func_70608_bn())
         {
-            int var7 = p_78718_1_.func_72798_a(MathHelper.func_76128_c(p_78718_4_.field_70165_t), MathHelper.func_76128_c(p_78718_4_.field_70163_u), MathHelper.func_76128_c(p_78718_4_.field_70161_v));
-
-            if (var7 == Block.field_71959_S.field_71990_ca)
+            int x = MathHelper.func_76128_c(p_78718_4_.field_70165_t);
+            int y = MathHelper.func_76128_c(p_78718_4_.field_70163_u);
+            int z = MathHelper.func_76128_c(p_78718_4_.field_70161_v);
+            Block block = Block.field_71973_m[p_78718_1_.func_72798_a(x, y, z)];
+            
+            if (block != null && block.isBed(p_78718_1_, x, y, z, p_78718_4_))
             {
-                int var8 = p_78718_1_.func_72805_g(MathHelper.func_76128_c(p_78718_4_.field_70165_t), MathHelper.func_76128_c(p_78718_4_.field_70163_u), MathHelper.func_76128_c(p_78718_4_.field_70161_v));
-                int var9 = var8 & 3;
+                int var9 = block.getBedDirection(p_78718_1_, x, y, z);
                 this.field_78735_i = (float)(var9 * 90 + 180);
                 this.field_78732_j = 0.0F;
             }
