--- a/net/minecraft/client/entity/EntityOtherPlayerMP.java
+++ b/net/minecraft/client/entity/EntityOtherPlayerMP.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.entity;
 
+import com.mojang.authlib.minecraft.MinecraftProfileTexture;
+import cpw.mods.fml.relauncher.FMLAuth;
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import net.minecraft.client.Minecraft;
@@ -32,7 +34,7 @@
 
         if (p_i3117_2_ != null && p_i3117_2_.length() > 0)
         {
-            this.field_70120_cr = "http://skins.minecraft.net/MinecraftSkins/" + StringUtils.func_76338_a(p_i3117_2_) + ".png";
+            this.field_70120_cr = FMLAuth.instance.getTextureURL(MinecraftProfileTexture.Type.SKIN);
         }
 
         this.field_70145_X = true;
@@ -62,7 +64,7 @@
 
     public void func_70059_ac()
     {
-        this.field_71084_cw = "http://skins.minecraft.net/MinecraftCloaks/" + StringUtils.func_76338_a(this.field_71092_bJ) + ".png";
+        this.field_71084_cw = FMLAuth.instance.getTextureURL(MinecraftProfileTexture.Type.CAPE);
         this.field_70119_cs = this.field_71084_cw;
     }
 
