--- a/net/minecraft/client/particle/EntityFireworkStarterFX.java
+++ b/net/minecraft/client/particle/EntityFireworkStarterFX.java
@@ -30,11 +30,11 @@
         {
             this.field_92039_az = p_i8002_15_.func_74761_m("Explosions");
 
-            if (this.field_92039_az.func_74745_c() == 0)
+            if (this.field_92039_az != null && this.field_92039_az.func_74745_c() == 0)
             {
                 this.field_92039_az = null;
             }
-            else
+            else if (this.field_92039_az != null)
             {
                 this.field_70547_e = this.field_92039_az.func_74745_c() * 2 - 1;
 
