--- a/net/minecraft/world/gen/ChunkProviderGenerate.java
+++ b/net/minecraft/world/gen/ChunkProviderGenerate.java
@@ -20,6 +20,12 @@
 import net.minecraft.world.gen.structure.MapGenStronghold;
 import net.minecraft.world.gen.structure.MapGenVillage;
 
+import static net.minecraftforge.event.terraingen.InitMapGenEvent.EventType.*;
+import static net.minecraftforge.event.terraingen.PopulateChunkEvent.Populate.EventType.*;
+import net.minecraftforge.common.*;
+import net.minecraftforge.event.Event.*;
+import net.minecraftforge.event.terraingen.*;
+
 public class ChunkProviderGenerate implements IChunkProvider
 {
     private Random field_73220_k;
@@ -48,6 +54,15 @@
     double[] field_73221_h;
     float[] field_73222_i;
     int[][] field_73219_j = new int[32][32];
+    
+    {
+        field_73226_t = TerrainGen.getModdedMapGen(field_73226_t, CAVE);
+        field_73225_u = (MapGenStronghold) TerrainGen.getModdedMapGen(field_73225_u, STRONGHOLD);
+        field_73224_v = (MapGenVillage) TerrainGen.getModdedMapGen(field_73224_v, VILLAGE);
+        field_73223_w = (MapGenMineshaft) TerrainGen.getModdedMapGen(field_73223_w, MINESHAFT);
+        field_73233_x = (MapGenScatteredFeature) TerrainGen.getModdedMapGen(field_73233_x, SCATTERED_FEATURE);
+        field_73232_y = TerrainGen.getModdedMapGen(field_73232_y, RAVINE);
+    }
 
     public ChunkProviderGenerate(World p_i3782_1_, long p_i3782_2_, boolean p_i3782_4_)
     {
@@ -61,6 +76,16 @@
         this.field_73214_a = new NoiseGeneratorOctaves(this.field_73220_k, 10);
         this.field_73212_b = new NoiseGeneratorOctaves(this.field_73220_k, 16);
         this.field_73213_c = new NoiseGeneratorOctaves(this.field_73220_k, 8);
+        
+        NoiseGeneratorOctaves[] noiseGens = {field_73217_l, field_73218_m, field_73215_n, field_73216_o, field_73214_a, field_73212_b, field_73213_c};
+        noiseGens = TerrainGen.getModdedNoiseGenerators(p_i3782_1_, this.field_73220_k, noiseGens);
+        this.field_73217_l = noiseGens[0];
+        this.field_73218_m = noiseGens[1];
+        this.field_73215_n = noiseGens[2];
+        this.field_73216_o = noiseGens[3];
+        this.field_73214_a = noiseGens[4];
+        this.field_73212_b = noiseGens[5];
+        this.field_73213_c = noiseGens[6];
     }
 
     public void func_73206_a(int p_73206_1_, int p_73206_2_, byte[] p_73206_3_)
@@ -139,6 +164,10 @@
 
     public void func_73207_a(int p_73207_1_, int p_73207_2_, byte[] p_73207_3_, BiomeGenBase[] p_73207_4_)
     {
+        ChunkProviderEvent.ReplaceBiomeBlocks event = new ChunkProviderEvent.ReplaceBiomeBlocks(this, p_73207_1_, p_73207_2_, p_73207_3_, p_73207_4_);
+        MinecraftForge.EVENT_BUS.post(event);
+        if (event.getResult() == Result.DENY) return;
+        
         byte var5 = 63;
         double var6 = 0.03125D;
         this.field_73227_s = this.field_73216_o.func_76304_a(this.field_73227_s, p_73207_1_ * 16, p_73207_2_ * 16, 0, 16, 16, 1, var6 * 2.0D, var6 * 2.0D, var6 * 2.0D);
@@ -263,6 +292,10 @@
 
     private double[] func_73205_a(double[] p_73205_1_, int p_73205_2_, int p_73205_3_, int p_73205_4_, int p_73205_5_, int p_73205_6_, int p_73205_7_)
     {
+        ChunkProviderEvent.InitNoiseField event = new ChunkProviderEvent.InitNoiseField(this, p_73205_1_, p_73205_2_, p_73205_3_, p_73205_4_, p_73205_5_, p_73205_6_, p_73205_7_);
+        MinecraftForge.EVENT_BUS.post(event);
+        if (event.getResult() == Result.DENY) return event.noisefield;
+        
         if (p_73205_1_ == null)
         {
             p_73205_1_ = new double[p_73205_5_ * p_73205_6_ * p_73205_7_];
@@ -425,6 +458,8 @@
         this.field_73220_k.setSeed((long)p_73153_2_ * var7 + (long)p_73153_3_ * var9 ^ this.field_73230_p.func_72905_C());
         boolean var11 = false;
 
+        MinecraftForge.EVENT_BUS.post(new PopulateChunkEvent.Pre(p_73153_1_, field_73230_p, field_73220_k, p_73153_2_, p_73153_3_, var11));
+        
         if (this.field_73229_q)
         {
             this.field_73223_w.func_75051_a(this.field_73230_p, this.field_73220_k, p_73153_2_, p_73153_3_);
@@ -437,7 +472,8 @@
         int var13;
         int var14;
 
-        if (!var11 && this.field_73220_k.nextInt(4) == 0)
+        if (TerrainGen.populate(p_73153_1_, field_73230_p, field_73220_k, p_73153_2_, p_73153_3_, var11, LAKE) && 
+                !var11 && this.field_73220_k.nextInt(4) == 0)
         {
             var12 = var4 + this.field_73220_k.nextInt(16) + 8;
             var13 = this.field_73220_k.nextInt(128);
@@ -445,7 +481,8 @@
             (new WorldGenLakes(Block.field_71943_B.field_71990_ca)).func_76484_a(this.field_73230_p, this.field_73220_k, var12, var13, var14);
         }
 
-        if (!var11 && this.field_73220_k.nextInt(8) == 0)
+        if (TerrainGen.populate(p_73153_1_, field_73230_p, field_73220_k, p_73153_2_, p_73153_3_, var11, LAVA) &&
+                !var11 && this.field_73220_k.nextInt(8) == 0)
         {
             var12 = var4 + this.field_73220_k.nextInt(16) + 8;
             var13 = this.field_73220_k.nextInt(this.field_73220_k.nextInt(120) + 8);
@@ -457,7 +494,8 @@
             }
         }
 
-        for (var12 = 0; var12 < 8; ++var12)
+        boolean doGen = TerrainGen.populate(p_73153_1_, field_73230_p, field_73220_k, p_73153_2_, p_73153_3_, var11, DUNGEON);
+        for (var12 = 0; doGen && var12 < 8; ++var12)
         {
             var13 = var4 + this.field_73220_k.nextInt(16) + 8;
             var14 = this.field_73220_k.nextInt(128);
@@ -474,7 +512,8 @@
         var4 += 8;
         var5 += 8;
 
-        for (var12 = 0; var12 < 16; ++var12)
+        doGen = TerrainGen.populate(p_73153_1_, field_73230_p, field_73220_k, p_73153_2_, p_73153_3_, var11, ICE);
+        for (var12 = 0; doGen && var12 < 16; ++var12)
         {
             for (var13 = 0; var13 < 16; ++var13)
             {
@@ -491,6 +530,8 @@
                 }
             }
         }
+        
+        MinecraftForge.EVENT_BUS.post(new PopulateChunkEvent.Post(p_73153_1_, field_73230_p, field_73220_k, p_73153_2_, p_73153_3_, var11));
 
         BlockSand.field_72192_a = false;
     }
