--- ../src-base/minecraft/net/minecraft/inventory/SlotFurnace.java
+++ ../src-work/minecraft/net/minecraft/inventory/SlotFurnace.java
@@ -1,5 +1,6 @@
 package net.minecraft.inventory;
 
+import cpw.mods.fml.common.registry.GameRegistry;
 import net.minecraft.entity.item.EntityXPOrb;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.Item;
@@ -53,7 +54,7 @@
         if (!this.field_75229_a.field_70170_p.field_72995_K)
         {
             int var2 = this.field_75228_b;
-            float var3 = FurnaceRecipes.func_77602_a().func_77601_c(p_75208_1_.field_77993_c);
+            float var3 = FurnaceRecipes.func_77602_a().getExperience(p_75208_1_);
             int var4;
 
             if (var3 == 0.0F)
@@ -81,6 +82,8 @@
         }
 
         this.field_75228_b = 0;
+
+        GameRegistry.onItemSmelted(field_75229_a, p_75208_1_);
 
         if (p_75208_1_.field_77993_c == Item.field_77703_o.field_77779_bT)
         {
